client/src/lib/actions/profile_setup.ts

```diff
@@ -21,63 +34,171 @@ export type ProfileSetupState = {
-    school,
-    avatar,
-  });
+//getProfileForUser gets profile for currently auth'd user
+export async function getProfileForUser() : Promise<Json|null>{
+  debug("start of getProfileForUser() function");
+  const user_id = await getUserId();
+  if(!user_id) {
+    debug("No user_id: GetUserId() returned null");
+    return null;
+  }
+  const supabase = createSupabaseServer();
+    
+  //query for profile by user_id
+  const { data, error } = await supabase
+    .from("profile")
+    .select('*')
+    .eq("user_id", user_id);
+
+    if (error) {
+      debug("ERROR: unable to fetch profile by userid");
+      debug("error: "+ error);
+      return null;
+    }
+
+    const profileDataStr = JSON.stringify(data, null, 2);
+    const profileJson = JSON.parse(profileDataStr);
+    
+    if (profileJson && profileJson[0]) {
+      return profileJson[0];
+    }
+    else {
+      return null
+    }
+} // end of getProfileForUser
+
+export async function getProfileByUserName(username: string) : Promise<Json|null>{
+
+  debug("start of getProfileByUserName() function");
+  const supabase = createSupabaseServer();
+    
+  //query for profile by username
+  const { data, error } = await supabase
+    .from("profile")
+    .select('*')
+    .eq("username", username);
+
+    if (error) {
+      debug("ERROR: retrieving profile by username: "+ error);
+      return null;
+    }
 
-  if (error) {
-    console.log("Error occurred creating profile:", error);
-    return { message: "Database Error" };
+    const profileDataStr = JSON.stringify(data, null, 2);
+    const profileJson = JSON.parse(profileDataStr);
+
+    if (profileJson && profileJson[0]) {
+      return profileJson[0];
+    }
+    else {
+      return null
+    }
+} // end of getProfileByUserName
+
+export async function createProfile( prevState: ProfileSetupState, formData: FormData ): Promise<ProfileSetupState> {
+  if (await userHasExistingProfile()) {
+    debug("USER has an existing profile.")
+    return {
+      errors: {
+        school: ["ERROR: User has an existing profile"],
+      } 
+    };
   }
+  else {
+    debug("USER doesn't have a profile");
+    // Validate form data with Zod schema
+    const validatedFields = ProfileSetupFormSchema.safeParse({
+      name: formData.get("name"),
+      username: formData.get("username"),
+      school: formData.get("school"),
+      avatar: formData.get("avatar"),
+    });
 
-  // Redirect to the dashboard after successful profile creation
-  redirect("/dashboard");
-}
+    if (!validatedFields.success) {
+      debug("ERROR: zod validation failed");
+      return { errors: validatedFields.error.flatten().fieldErrors };
+    }
+    else {
+      debug("Zod validation success, checking unique username");
+      const supabase = createSupabaseServer();
+      const { data: {user}, error: userError } =  await supabase.auth.getUser();
+      if (userError) {
+          return { 
+            errors: { school: ["Error: No user"] }
```

Comment:
Users with no account created shouldn't also be able to access this page. Handle redirecting them to a 404 via the middleware if they try to manually go to this page so you don't have to handle the check if they have an account when they create profile. You should assume that all users going to this page have not created a profile yet